cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/DataprocUtils.java

```diff
@@ -292,6 +302,58 @@ private static String getMetadata(String resource) throws IOException {
     }
   }
 
+  /**
+   * Recursively deletes all the contents of the directory and the directory itself.
+   */
+  public static void deleteDirectoryContents(@Nullable File file) {
+    if (file == null) {
+      return;
+    }
+
+    if (file.isDirectory()) {
+      File[] entries = file.listFiles();
+      if (entries != null) {
+        for (File entry : entries) {
+          deleteDirectoryContents(entry);
+        }
+      }
+    }
+    if (!file.delete()) {
+      LOG.warn("Failed to delete file {}.", file);
+    }
+  }
+
+  /**
+   * Recursively deletes all the contents of the directory and the directory itself with retries.
+   */
+  public static synchronized void deleteDirectoryWithRetries(@Nullable File file)  {
+    ExponentialBackOff backOff = new ExponentialBackOff.Builder()
+      .setInitialIntervalMillis(MIN_WAIT_TIME_MILLISECOND)
+      .setMaxIntervalMillis(MAX_WAIT_TIME_MILLISECOND).build();
+
+    Exception exception = null;
+    int counter = 0;
+    while (counter < NUMBER_OF_RETRIES) {
+      counter++;
+
+      try {
+        deleteDirectoryContents(file);
+        return;
+      } catch (Exception e) {
+        // exception does not get logged since it might get too chatty.
+        exception = e;
+      }
+
+      try {
+        Thread.sleep(backOff.nextBackOffMillis());
+      } catch (InterruptedException | IOException e) {
+        exception = e;
+        break;
+      }
+    }
+    throw new RuntimeException(String.format("Unable to delete directory %s.", file), exception);
```

Comment:
nit: unable to delete local cache directory %s for twill.jar and launcher.jar

Reply:
Since the method is in util class, didn't wanted to add information about `twill and launcher jar and local cache`, will send the error message as an input param.